packages/server-boilerplate/src/models/types.ts

```diff
@@ -84,7 +84,19 @@ export type Joined<T, U extends string> = {
   [field in keyof T as field extends string ? `${U}.${field}` : never]: T[field];
 };
 
-export type Model<BaseModel extends DatabaseModel, Fields, Type extends DatabaseType> = {
-  find: (filter: DbFilter<Fields>) => Promise<Type[]>;
-  findOne: (filter: DbFilter<Fields>) => Promise<Type>;
-} & BaseModel;
+export type QueryOptions = {
+  limit?: number;
+  offset?: number;
+  sort?: string[];
+};
+
+type BaseModelOverrides<Fields = unknown, Type = unknown> = {
+  find: (filter: DbFilter<Fields>, customQueryOptions?: QueryOptions) => Promise<Type[]>;
+  findOne: (filter: DbFilter<Fields>, customQueryOptions?: QueryOptions) => Promise<Type>;
+};
+
+export type Model<BaseModel extends DatabaseModel, Fields, Type extends DatabaseType> = Omit<
+  BaseModel,
+  keyof BaseModelOverrides
+> &
+  BaseModelOverrides<Fields, Type>;
```

Comment:
I'm a bit worried when I add these super complicated types 😓 They make life a lot easier when they're working, but understanding them is really complicated. I think as we grow more mature as a typescript codebase we'll find out how best to strike the balance.

The reason for this one is to basically force using the 'overloaded' types for the model functions (`findOne`, `find`,etc.). I had to do this since the ones in base type use `any` a lot, and it was kind of messing with the types in the actual typescript packages...

Reply:
Ah yep that makes sense